### PR TITLE
[7.x] Fix array eloquent casting

### LINKED ISSUE #32117: Ambiguities in array eloquent casting
- Laravel Version: 6.x/7.x
- PHP Version: 7.2. Actual on 7.3, 7.4
- Database Driver & Version: MySQL 5.6/MariaDB 10

### Description:

Array casting not working as I'm expecting.
I expect that invalid json string to be converted to an array or an exception to be thrown. But invalid json converts to string.

Invalid json `string(2) """"` converted to empty string.

This happens because json_decode convert some invalid json to ambiguities value. Example:
```php
$string = '""';
var_dump(json_decode($string, true)); // string(0) ""
```

Json string to array casting uses `json_decode` without any checks.
https://github.com/laravel/framework/blob/7.x/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L809

### Steps To Reproduce:

* Create some database entity

Migration:
```php
Schema::create('foo', function (Blueprint $table) {
    $table->increments('id');
    $table->string('payload');
});
```

Model:
```php
class Foo extends Model
{
    public $timestamps = false; 

    protected $casts = [
        'bar' => 'array'
    ];
}

```

* Insert new row to database with invalid json:

```sql
INSERT INTO `foo` (`bar`) VALUES ('""');
```

* Find element and read bar property

```
$foo = App\Foo::find(1);

var_dump($foo->bar); // Expects array but given empty string
```

### PR DESCRIPTION
**Pull request by et-nik**
This PR fixes #32117 issue

Also this issue is actual on 6.x and 8.x vesions

### CODE REVIEW

**Comment by taylorotwell** [comment]
Why are you storing an empty string in a database column that is supposed to contain array data?

**Comment by et-nik** [comment]
I'm not storing. I think that this is pointless.
But I left this opportunity in order not to break backward compatibility.

**Comment by et-nik** [comment]
I want to change only `fromJson` method:
```php
    public function fromJson($value, $asObject = false)
    {
        $decoded = json_decode($value, ! $asObject);

        if ($asObject) {
            return is_object($decoded) ? $decoded : new stdClass;
        }

        return is_array($decoded) ? $decoded : [];
    }
```

But some tests fails.

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 1]
```diff
@@ -809,6 +811,30 @@ public function fromJson($value, $asObject = false)
...
+    protected function fromArrayJson($value)
+    {
+        $decoded = $this->fromJson($value, false);
+
+        return is_array($decoded) ? $decoded : [];
+    }
+
+    /**
+     * Decode the given JSON into an object.
+     * @param $value
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: invalid syntax

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 2]
```diff
@@ -809,6 +811,30 @@ public function fromJson($value, $asObject = false)
         return json_decode($value, ! $asObject);
     }
 
+    /**
+     * Decode the given JSON into an array.
+     * @param $value
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: invalid syntax

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by taylorotwell** [comment]
Don't store empty strings in the database for array / json columns and this won't happen.

**Comment by et-nik** [comment]
@taylorotwell Casts allow do it. Why?
If I add 'array' cast I doesn't expect string or bool value. But this is happening.

Model:
```
class SomeModel extends Eloquent
{
    /**
     * @var string
     */
    protected $table = 'some_table';

    /**
     * @var array
     */
    protected $guarded = [];

    /**
     * @var array
     */
    protected $casts = [
        'array_attribute'   => 'array',
        'object_attribute'  => 'object',
        'json_attribute'    => 'json',
    ];
}
```

Laravel save string json.
```
$someModel = SomeModel::find(1);
$someModel->array_attribute = 'string';
$someModel->save();

$someModel2 = SomeModel::find(1);
echo $someModel2->array_attribute; // 'string'. Why it is string??? I expect array
```

I think you should not be allowed to keep this.

**Pull request closed without merging**